Refactor istioctl analyze and analysis messages

niwatorrri opened the pull request:
Move message specific methods from istioctl analyze to `galley/pkg/config/analysis/diag` package. This is a first step toward [unifying istioctl command output](https://docs.google.com/document/d/1_m3PRHjbSFy6NAOk4C4CvCu93q-QKRsW7WaiaY9Z1DQ/edit#heading=h.xw1gqgyqs5b).

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience
[ ] Developer Infrastructure

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | f480751a4383c1974c8f802b23a705ecf1bc0ba3 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/928) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | f480751a4383c1974c8f802b23a705ecf1bc0ba3 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/928) | `/test release-notes_istio`
integ-pilot-k8s-tests_istio | f480751a4383c1974c8f802b23a705ecf1bc0ba3 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17344) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented:
For [unifying istioctl command output](https://docs.google.com/document/d/1_m3PRHjbSFy6NAOk4C4CvCu93q-QKRsW7WaiaY9Z1DQ/edit#heading=h.xw1gqgyqs5b), I intend to break this into 2-3 PRs. This PR focuses on methods for Messages and istioctl analyze. Precheck and verify-install will be done in followup PRs.

cc @nmittler @davidhauck @maasen @jasonwzm

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | f480751a4383c1974c8f802b23a705ecf1bc0ba3 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17344) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 8751990c22a9309ae52729cdaacd27e4535957ce | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/935) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 8751990c22a9309ae52729cdaacd27e4535957ce | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/935) | `/test release-notes_istio`
integ-pilot-k8s-tests_istio | 8751990c22a9309ae52729cdaacd27e4535957ce | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17351) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 8751990c22a9309ae52729cdaacd27e4535957ce | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17351) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/937) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented:
/release-notes-none

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 8751990c22a9309ae52729cdaacd27e4535957ce | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17351) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/937) | `/test release-notes_istio`
unit-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17841) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented:
/retest

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 8751990c22a9309ae52729cdaacd27e4535957ce | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17351) | `/test integ-pilot-k8s-tests_istio`
unit-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17841) | `/test unit-tests_istio`
release-notes_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/938) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17841) | `/test unit-tests_istio`
release-notes_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/938) | `/test release-notes_istio`
integ-pilot-k8s-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17353) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/938) | `/test release-notes_istio`
integ-pilot-k8s-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17353) | `/test integ-pilot-k8s-tests_istio`
unit-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17842) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17353) | `/test integ-pilot-k8s-tests_istio`
unit-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17842) | `/test unit-tests_istio`
release-notes_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/957) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17353) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/957) | `/test release-notes_istio`
unit-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17861) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 5708b294b46fc117466f70047cb16a12790b3055 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17353) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/957) | `/test release-notes_istio`
unit-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17861) | `/test unit-tests_istio`
integ-multicluster-k8s-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-multicluster-k8s-tests_istio/7883) | `/test integ-multicluster-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/957) | `/test release-notes_istio`
unit-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17861) | `/test unit-tests_istio`
integ-multicluster-k8s-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-multicluster-k8s-tests_istio/7883) | `/test integ-multicluster-k8s-tests_istio`
integ-pilot-k8s-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17373) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

jasonwzm commented:
/cc @maasen

jasonwzm commented on galley/pkg/config/analysis/diag/messages.go:
```diff
@@ -14,7 +14,32 @@
+)
+
+func init() {
+	sort.Strings(MsgOutputFormatKeys)
```
Why do we need to do this?

niwatorrri replied:
It was on line 105 in the original [code](https://github.com/istio/istio/pull/25996/files/0ad7cc516a4fd09163cd48033a23be4407f140fb#diff-b3e374e4ac15f6818ab702c5172ce519L105). I kept it as it was but sure we can remove this.

zerobfd replied:
You can use the blame interface in the original file to see who added a particular line at what commit/pull. This was added in #24212 and is needed.

zerobfd commented on galley/pkg/config/analysis/diag/level.go:
```diff
@@ -68,3 +69,36 @@ func GetUppercaseStringToLevelMap() map[string]Level {
 }
+
+// MessageThreshold is a wrapper around Level to use as cobra command line arguments
+type MessageThreshold struct {
+	Level
+}
```
1) I don't really understand why we need another struct for this. A Level _is_ a threshold, and there's no additional type safety gained here.

2) FYI: You can just ```type MessageThreshold Level```

niwatorrri replied:
The MessageThreshold struct is one of the codes which I kept as was, but this struct is only used by istioctl analyze for now so yeah I could merge it with Level.

zerobfd commented on galley/pkg/config/analysis/diag/level.go:
```diff
@@ -68,3 +69,36 @@ func GetUppercaseStringToLevelMap() map[string]Level {
+	Level
+}
+
+// String satisfies interface pflag.Value
```
I think this satisfies a different interface.

niwatorrri replied:
I believe this is correct, according to what I found [here](https://godoc.org/github.com/spf13/pflag#Value).

zerobfd replied:
Huh, that's really weird. You're right--I assumed that pflag would just use the Stringer interface instead of specifying a String method. Mea culpa.

zerobfd commented on galley/pkg/config/analysis/diag/level.go:
```diff
@@ -68,3 +69,36 @@ func GetUppercaseStringToLevelMap() map[string]Level {
+}
+
+// Set satisfies interface pflag.Value
+func (m *MessageThreshold) Set(s string) error {
+	l, err := levelFromString(s)
+	if err != nil {
+		return err
+	}
+	m.Level = l
+	return nil
+}
```
I don't think the if statement is necessary, this seems to be the same:

```go
func (m *MessageThreshold) Set(s string) error {
	l, err := levelFromString(s)
	m.Level = l
	return err
}
```

zerobfd commented on galley/pkg/config/analysis/diag/level.go:
```diff
@@ -68,3 +69,36 @@ func GetUppercaseStringToLevelMap() map[string]Level {
+func levelFromString(s string) (Level, error) {
+	val, ok := GetUppercaseStringToLevelMap()[strings.ToUpper(s)]
+	if !ok {
+		return Level{}, fmt.Errorf("%q not a valid option, please choose from: %v", s, GetAllLevelStrings())
```
Error strings should just state what's wrong, not be worded as if it were always to be printed. See https://github.com/golang/go/wiki/CodeReviewComments#error-strings for some rationale.

In general the top level caller is responsible for formatting the string further as in the link's example: 
```log.Printf("Reading %s: %v", filename, err)```

In this specific instance you can set validation functions with Cobra so you can do validation at the CLI.

zerobfd commented on galley/pkg/config/analysis/diag/level.go:
```diff
@@ -68,3 +69,36 @@ func GetUppercaseStringToLevelMap() map[string]Level {
+	return nil
+}
+
+func levelFromString(s string) (Level, error) {
```
"FromString" is generally "Parse" in golang, so "parseLevel" makes more sense to me.

niwatorrri commented:
@zerobfd Thanks for the review, Clay! Some followup comments before I redo some code:

- The reason why I moved them from istioctl analyze into messages is that: I'd like to make other istioctl commands (precheck and verify-install) produce messages in the same way as istioctl analyze does, and this move intends to provide a uniform way to format all these messages across different commands. If it sounds better we can separate these format options into a MessageFormatter or something.
- I assume galley is already deprecated? If so, I agree that we could move these to somewhere like `istioctl/cmd/diag`. Other code, e.g., the `galley/pkg/config/analysis/msg` package, might also be better moved to a similar location.
- I kept most of the code and comments I moved as how they originally were. There might be some reasons why these were designed in a particular way, but if it's harmless we should improve them.
- I would be all ears if you have recommended practices for easier transition to using analysis message API later. I cannot think of apparent disadvantages of the current approach right now.

zerobfd commented:
> @zerobfd Thanks for the review, Clay! Some followup comments before I go change some code:
> 
> * The reason why I moved them from istioctl analyze into messages is that: I'd like to make other istioctl commands (precheck and verify-install) produce messages in the same way as istioctl analyze does, and this move intends to provide a uniform way to format all these messages across different commands. If it sounds better we can separate these format options into a MessageFormatter or something.

I'm all for code re-use and agree that we should standardize structured output for all of the istioctl commands. I'd be extremely happy if you just took all the code as-is and put it into a MessageFormatter. (Side note: not sure if this was what you were thinking, but I don't think we need a new struct for this--we can just use plain functions.)

> * I assume galley is already deprecated? If so, I agree that we could move these to somewhere like `istioctl/cmd/diag`. Other code, e.g., the `galley/pkg/config/analysis/msg` package, might also be better moved to a similar location.

Yes and I agree. This formatting code clearly fits under istioctl. A lot of the other code under /galley is used in other places, so figuring out 1) where to move it and 2) all of the dependencies that need to be updated is a lot of work. If you're interested in tackling that I'd love to support you, but not in this PR.

> * I kept most of the code and comments I moved as how they originally were. There might be some reasons why these were designed in a particular way, but if it's harmless we should improve them.

I agree in general, but the devil is in the details. "Seems harmless" is very different than "verified with the original author/design doc that it can be improved safely." If there are places that you want to improve during the move feel free to flag them and someone can chime in on whether it's safe.

> * I would be all ears if you have recommended practices for easier transition to using analysis message API later. I cannot think of apparent disadvantages of the current approach right now.

The big thing is just to not tie the formatting logic to the Message class. Using the message API later means using protocol buffers, and generally you want to keep as much logic as possible out of proto structs and leave them as just data containers that can handle some basic typing and serialization.

Thanks for slogging through all of this, I think standardizing the CLI is really important and I appreciate your work to move it in the right direction.

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17861) | `/test unit-tests_istio`
integ-multicluster-k8s-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-multicluster-k8s-tests_istio/7883) | `/test integ-multicluster-k8s-tests_istio`
integ-pilot-k8s-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17373) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 4f6091f9d1eb61bdd4f9d7d58a35f148dd2997c2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1115) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/17861) | `/test unit-tests_istio`
integ-multicluster-k8s-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-multicluster-k8s-tests_istio/7883) | `/test integ-multicluster-k8s-tests_istio`
integ-pilot-k8s-tests_istio | 0ad7cc516a4fd09163cd48033a23be4407f140fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17373) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 4f6091f9d1eb61bdd4f9d7d58a35f148dd2997c2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1115) | `/test release-notes_istio`
integ-distroless-k8s-tests_istio | 4f6091f9d1eb61bdd4f9d7d58a35f148dd2997c2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-distroless-k8s-tests_istio/10641) | `/test integ-distroless-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | 4f6091f9d1eb61bdd4f9d7d58a35f148dd2997c2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-distroless-k8s-tests_istio/10641) | `/test integ-distroless-k8s-tests_istio`
release-notes_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1125) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | 4f6091f9d1eb61bdd4f9d7d58a35f148dd2997c2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-distroless-k8s-tests_istio/10641) | `/test integ-distroless-k8s-tests_istio`
release-notes_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1125) | `/test release-notes_istio`
release-test_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-test_istio/17234) | `/test release-test_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | 4f6091f9d1eb61bdd4f9d7d58a35f148dd2997c2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-distroless-k8s-tests_istio/10641) | `/test integ-distroless-k8s-tests_istio`
release-notes_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1125) | `/test release-notes_istio`
release-test_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-test_istio/17234) | `/test release-test_istio`
integ-pilot-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17556) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | 4f6091f9d1eb61bdd4f9d7d58a35f148dd2997c2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-distroless-k8s-tests_istio/10641) | `/test integ-distroless-k8s-tests_istio`
release-notes_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1125) | `/test release-notes_istio`
release-test_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-test_istio/17234) | `/test release-test_istio`
integ-pilot-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17556) | `/test integ-pilot-k8s-tests_istio`
integ-ipv6-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-ipv6-k8s-tests_istio/3640) | `/test integ-ipv6-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1125) | `/test release-notes_istio`
release-test_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-test_istio/17234) | `/test release-test_istio`
integ-pilot-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17556) | `/test integ-pilot-k8s-tests_istio`
integ-ipv6-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-ipv6-k8s-tests_istio/3640) | `/test integ-ipv6-k8s-tests_istio`
integ-distroless-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-distroless-k8s-tests_istio/10653) | `/test integ-distroless-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1125) | `/test release-notes_istio`
release-test_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-test_istio/17234) | `/test release-test_istio`
integ-pilot-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17556) | `/test integ-pilot-k8s-tests_istio`
integ-ipv6-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-ipv6-k8s-tests_istio/3640) | `/test integ-ipv6-k8s-tests_istio`
integ-distroless-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-distroless-k8s-tests_istio/10653) | `/test integ-distroless-k8s-tests_istio`
integ-operator-controller-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-operator-controller-tests_istio/5799) | `/test integ-operator-controller-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1125) | `/test release-notes_istio`
release-test_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-test_istio/17234) | `/test release-test_istio`
integ-pilot-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17556) | `/test integ-pilot-k8s-tests_istio`
integ-ipv6-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-ipv6-k8s-tests_istio/3640) | `/test integ-ipv6-k8s-tests_istio`
integ-distroless-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-distroless-k8s-tests_istio/10653) | `/test integ-distroless-k8s-tests_istio`
integ-operator-controller-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-operator-controller-tests_istio/5799) | `/test integ-operator-controller-tests_istio`
unit-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/18037) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-test_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-test_istio/17234) | `/test release-test_istio`
integ-pilot-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17556) | `/test integ-pilot-k8s-tests_istio`
integ-ipv6-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-ipv6-k8s-tests_istio/3640) | `/test integ-ipv6-k8s-tests_istio`
integ-distroless-k8s-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-distroless-k8s-tests_istio/10653) | `/test integ-distroless-k8s-tests_istio`
integ-operator-controller-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-operator-controller-tests_istio/5799) | `/test integ-operator-controller-tests_istio`
unit-tests_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/18037) | `/test unit-tests_istio`
release-notes_istio | 3e4e68adb5cf26638226bc177787f9e9d88ea2dc | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1126) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-test_istio | 6df14d27303c098e5ed9a0fbec94cb106d8311f1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-test_istio/17234) | `/test release-test_istio`
release-notes_istio | 3e4e68adb5cf26638226bc177787f9e9d88ea2dc | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1126) | `/test release-notes_istio`
integ-pilot-k8s-tests_istio | 3e4e68adb5cf26638226bc177787f9e9d88ea2dc | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17557) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented:
/test integ-pilot-k8s-tests_istio

zerobfd commented on galley/pkg/config/analysis/diag/level.go:
```diff
@@ -68,3 +69,19 @@ func GetUppercaseStringToLevelMap() map[string]Level {
 }
+
+// Type satisfies interface pflag.Value
+func (l *Level) Type() string {
```
>> The MessageThreshold struct is one of the codes that I kept as was, but this struct is only used by istioctl analyze for now so yeah I could merge it with Level.

Ok, now I get what you meant by that--you mean the private struct that was in analyze.go. I'll address this more in the review comment.

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | bcfe9a8fc3c974cedcadd3417d1347c0cf8f5a24 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1190) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | bcfe9a8fc3c974cedcadd3417d1347c0cf8f5a24 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1190) | `/test release-notes_istio`
lint_istio | bcfe9a8fc3c974cedcadd3417d1347c0cf8f5a24 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/lint_istio/17433) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | bcfe9a8fc3c974cedcadd3417d1347c0cf8f5a24 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1190) | `/test release-notes_istio`
lint_istio | bcfe9a8fc3c974cedcadd3417d1347c0cf8f5a24 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/lint_istio/17433) | `/test lint_istio`
integ-pilot-k8s-tests_istio | bcfe9a8fc3c974cedcadd3417d1347c0cf8f5a24 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17634) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | bcfe9a8fc3c974cedcadd3417d1347c0cf8f5a24 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/lint_istio/17433) | `/test lint_istio`
integ-pilot-k8s-tests_istio | bcfe9a8fc3c974cedcadd3417d1347c0cf8f5a24 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17634) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 59a35ee0132816668249b4b419a90305a259dbff | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1191) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | bcfe9a8fc3c974cedcadd3417d1347c0cf8f5a24 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-pilot-k8s-tests_istio/17634) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 59a35ee0132816668249b4b419a90305a259dbff | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1191) | `/test release-notes_istio`
unit-tests_istio | 59a35ee0132816668249b4b419a90305a259dbff | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/18127) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 59a35ee0132816668249b4b419a90305a259dbff | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/unit-tests_istio/18127) | `/test unit-tests_istio`
release-notes_istio | 6b2287e33011271aee32e73263d6fdcc5bc6c9ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1192) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented on galley/pkg/config/analysis/diag/formatter.go:
```diff
@@ -0,0 +1,58 @@
+	for _, m := range ms {
+		logOutput = append(logOutput, m.Render(colorize))
+	}
+	return strings.Join(logOutput, "\n"), nil
```
I'd stick with strings.Join---there should not be a newline after the last message as enforced by the integration test, which requires strings.Builder to do more checking on the Messages.

niwatorrri commented on galley/pkg/config/analysis/diag/messages.go:
```diff
@@ -62,3 +82,22 @@ func (ms *Messages) SortedDedupedCopy() Messages {
+}
+
+// Filter only keeps messages at or above the specified output level
+func (ms *Messages) Filter(outputLevel Level) Messages {
```
Good point! I changed it to "FilterOutLowerThan" which hopefully should sound better.

niwatorrri commented:
@zerobfd Thanks for the detailed review! I really appreciate it. I reorganized the code as you suggested: in the `istio/istioctl/pkg/util/formatting`, I put format related options into formatter.go and MessageThreshold into a msg_threshold.go. I kind of preferred `type MessageThreshold struct { diag.Level }` to `type MessageThreshold diag.Level`, as I think when converting between the types `threshold.Level` might read better than `diag.Level(threshold)`.

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | e75d8b5beb92025913588de66df05af47873c559 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1221) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

zerobfd commented on galley/pkg/config/analysis/diag/messages_test.go:
```diff
@@ -114,11 +112,74 @@ func TestMessages_SortedCopy(t *testing.T) {
+		"B",
+	)
+	secondMsg := NewMessage(
+		NewMessageType(Info, "A1", "Template: %q"),
+		MockResource("B"),
+		"B",
+	)
+	thirdMsg := NewMessage(
+		NewMessageType(Warning, "C1", "Template: %q"),
+		MockResource("B"),
+		"B",
+	)
+
+	msgs := Messages{firstMsg, secondMsg, thirdMsg}
+	filteredMsgs := msgs.FilterOutLowerThan(Warning)
+	expectedMsgs := Messages{firstMsg, thirdMsg}
+
+	g.Expect(filteredMsgs).To(Equal(expectedMsgs))
+}
+
+func TestMessages_FilterOutAll(t *testing.T) {
+	g := NewWithT(t)
+
+	firstMsg := NewMessage(
+		NewMessageType(Info, "A1", "Template: %q"),
+		MockResource("B"),
+		"B",
+	)
+	secondMsg := NewMessage(
+		NewMessageType(Warning, "C1", "Template: %q"),
+		MockResource("B"),
+		"B",
+	)
+
+	msgs := Messages{firstMsg, secondMsg}
+	filteredMsgs := msgs.FilterOutLowerThan(Error)
+	expectedMsgs := Messages{}
+
+	g.Expect(filteredMsgs).To(Equal(expectedMsgs))
 }
```
Not necessary in this PR if you don't want to, but I think this would be a good candidate for table driven tests. See https://github.com/golang/go/wiki/TableDrivenTests

zerobfd commented on istioctl/cmd/analyze.go:
```diff
@@ -250,63 +227,43 @@ istioctl analyze -L
-					m.DocRef = "istioctl-analyze"
-					outputMessages = append(outputMessages, m)
-				}
+			// Append a ref arg to the doc URL, and filter outputMessages by specified level
+			outputMessages := result.Messages.SetDocRef("istioctl-analyze").FilterOutLowerThan(outputThreshold.Level)
```
Now that the function names are there I think this is pretty understandable without the comment.

zerobfd commented on istioctl/pkg/util/formatting/formatter_test.go:
```diff
@@ -0,0 +1,139 @@
+
+	. "github.com/onsi/gomega"
+
+	. "istio.io/istio/galley/pkg/config/analysis/diag"
```
Gomega afaik requires the dot import, but we should avoid it other places: https://github.com/golang/go/wiki/CodeReviewComments#import-dot

zerobfd commented on istioctl/pkg/util/formatting/formatter_test.go:
```diff
@@ -0,0 +1,139 @@
+	. "istio.io/istio/galley/pkg/config/analysis/diag"
+)
+
+func TestFormatter_RenderWithColor(t *testing.T) {
+	g := NewWithT(t)
+	mt := NewMessageType(Error, "IST-0042", "Cheese type not found: %q")
+	m := NewMessage(mt, nil, "Feta")
+
+	g.Expect(render(m, true)).To(Equal(
+		colorPrefixes[Error] + "Error\033[0m [IST-0042] Cheese type not found: \"Feta\""))
+}
```
Nit: ideally we can test all of the functionality of a class without touching non-exported pieces. Here if we want to check the color, I'd suggest checking it via Print(..., true) instead of poking directly at render() and colorPrefixes.

This serves two purposes.
1) It makes the tests more resistant to refactoring. Generally the public API doesn't change as much as the internals of a package, so if all of the tests go through exported methods there's less chance that changing the package means you also have to change a test.
2) It ensures that all of the code can actually get exercised with normal use. If there's not a way to test everything in the package without touching non-exported variables/functions, it's normally a sign that things might not be working as intended.

If you want to give it a try, I'd suggest changing the package to "formatting_test" instead of "formatting" so Go won't let you touch the internals. It's a small thing so I won't block acceptance on this.

Also, I think this entire file would also be a good candidate for table driven tests.

zerobfd commented on istioctl/pkg/util/formatting/formatter.go:
```diff
@@ -0,0 +1,111 @@
+	return string(yamlOutput), err
+}
+
+// Formatting options for Message
+var (
+	colorPrefixes = map[diag.Level]string{
+		diag.Info:    "",           // no special color for info messages
+		diag.Warning: "\033[33m",   // yellow
+		diag.Error:   "\033[1;31m", // bold red
+	}
+)
```
This looks like it's only used in the colorPrefix function, so it might be worth moving it there. (It's also used in the test file but I addressed that in the file itself.) I don't think making it package-global has any benefits, but I might be wrong.

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 70db8fe3784412e258f10226660638edac65f24d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/release-notes_istio/1248) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented:
@jasonwzm Looking to your approval. Also I can't find where to add a release-notes-none label. Do I not have enough permissions to do that?

jasonwzm commented:
/retest

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 70db8fe3784412e258f10226660638edac65f24d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/lint_istio/17556) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 70db8fe3784412e258f10226660638edac65f24d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/lint_istio/17556) | `/test lint_istio`
integ-security-k8s-tests_istio | 70db8fe3784412e258f10226660638edac65f24d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-security-k8s-tests_istio/17918) | `/test integ-security-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-k8s-tests_istio | 70db8fe3784412e258f10226660638edac65f24d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-security-k8s-tests_istio/17918) | `/test integ-security-k8s-tests_istio`
integ-multicluster-k8s-tests_istio | 3e25a16b0b0a320c9cd2b64a635918bfbf8439db | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25996/integ-multicluster-k8s-tests_istio/8247) | `/test integ-multicluster-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented:
/test integ-multicluster-k8s-tests_istio
